signature/jws/jws.go

```diff
@@ -0,0 +1,254 @@
+	// We unmarshal twice: once into a value of type jwsProtectedHeader and once into a value of type jwsProtectedHeader.ExtendedAttributes(map[string]interface{})
+	// and removing the keys are already been defined in jwsProtectedHeader.
+	var protected jwsProtectedHeader
+	if err = json.Unmarshal(rawProtected, &protected); err != nil {
+		return nil, &signature.InvalidSignatureError{
+			Msg: fmt.Sprintf("jws envelope protected header can't be decoded: %s", err.Error())}
+	}
+	if err = json.Unmarshal(rawProtected, &protected.ExtendedAttributes); err != nil {
+		return nil, &signature.InvalidSignatureError{
+			Msg: fmt.Sprintf("jws envelope protected header can't be decoded: %s", err.Error())}
+	}
+
+	// delete attributes that are already defined in jwsProtectedHeader.
+	for _, headerKey := range headerKeys {
+		delete(protected.ExtendedAttributes, headerKey)
+	}
+	return &protected, nil
+}
+
+func populateProtectedHeaders(protectedHeader *jwsProtectedHeader, signerInfo *signature.SignerInfo) error {
+	err := validateProtectedHeaders(protectedHeader)
+	if err != nil {
+		return err
+	}
+
+	if signerInfo.SignatureAlgorithm, err = getSignatureAlgorithm(protectedHeader.Algorithm); err != nil {
+		return err
+	}
+
+	signerInfo.SignedAttributes.ExtendedAttributes = getExtendedAttributes(protectedHeader.ExtendedAttributes, protectedHeader.Critical)
+	signerInfo.SignedAttributes.SigningScheme = protectedHeader.SigningScheme
+	if protectedHeader.Expiry != nil {
+		signerInfo.SignedAttributes.Expiry = *protectedHeader.Expiry
+	}
+	switch protectedHeader.SigningScheme {
+	case signature.SigningSchemeX509:
+		if protectedHeader.SigningTime != nil {
+			signerInfo.SignedAttributes.SigningTime = *protectedHeader.SigningTime
+		}
+	case signature.SigningSchemeX509SigningAuthority:
+		if protectedHeader.AuthenticSigningTime != nil {
+			signerInfo.SignedAttributes.SigningTime = *protectedHeader.AuthenticSigningTime
+		}
+	}
+	return nil
+}
+
+func validateProtectedHeaders(protectedHeader *jwsProtectedHeader) error {
+	// validate headers that should not be present as per signing schemes
+	switch protectedHeader.SigningScheme {
+	case signature.SigningSchemeX509:
+		if protectedHeader.AuthenticSigningTime != nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must not be present for %s signing scheme", headerKeyAuthenticSigningTime, signature.SigningSchemeX509)}
+		}
+	case signature.SigningSchemeX509SigningAuthority:
+		if protectedHeader.SigningTime != nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must not be present for %s signing scheme", headerKeySigningTime, signature.SigningSchemeX509SigningAuthority)}
+		}
+		if protectedHeader.AuthenticSigningTime == nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must be present for %s signing scheme", headerKeyAuthenticSigningTime, signature.SigningSchemeX509)}
+		}
+	}
```

Comment:
Should we check if SigningScheme is empty or unsupported?

Reply:
Added a default case.